Migrate bluetooth/host/buf.c UT to use FFF

ahmedmoheb-nordic opened the pull request:
WIP: Please check the modifications for the following files

- tests/bluetooth/host/buf/bt_buf_get_rx/src/main.c
- tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c

I added some behavior verification functions at tests/bluetooth/host/buf/mocks/buf_help_utils.c

ahmedmoheb-nordic committed:
tests: bluetooth: Add unit test projects for /bluetooth/host/buf.c

Unit testing for /bluetooth/host/buf.c APIs using ZTest.
Module APIs tested by UT projects:
- bt_buf_get_cmd_complete()
- bt_buf_get_evt()
- bt_buf_get_rx()
- bt_buf_set_type()
- bt_buf_get_type()

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
tests: bluetooth: host: Migrate bt_buf_get_rx() UT to use FFF

Use FFF testing framework instead of ZTest mocking.

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
tests: bluetooth: host: Migrate bt_buf_get_cmd_complete() UT to use FFF

Use FFF testing framework instead of ZTest mocking.

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic commented:
@jori-nordic @kruithofa please review

ahmedmoheb-nordic committed:
tests: bluetooth: host: Migrate bt_buf_get_evt() UT to use FFF

Use FFF testing framework instead of ZTest mocking.

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

jori-nordic commented on tests/bluetooth/host/buf/mocks/buf_help_utils.c:
```diff
@@ -0,0 +1,54 @@
+#include "net_buf.h"
+#include "buf_help_utils.h"
+
+void validate_net_buf_alloc_called_behaviour(struct net_buf_pool *pool, k_timeout_t *timeout)
```
I would rename that one: `void expect_single_call_net_buf_alloc(...)` (and the others using the same style)

jori-nordic commented on tests/bluetooth/host/buf/mocks/buf_help_utils.c:
```diff
@@ -0,0 +1,54 @@
+			  "'%s()' was called with incorrect '%s' value", func_name, "timeout");
+}
+
+void validate_net_buf_alloc_not_called_behaviour(void)
```
`void expect_not_called_net_buf_alloc(void)`

jori-nordic commented on tests/bluetooth/host/buf/mocks/net_buf.c:
```diff
@@ -7,50 +7,32 @@
-{
-	ztest_check_expected_value(value);
-}
+/* struct net_buf *net_buf_alloc_fixed(struct net_buf_pool *pool, k_timeout_t timeout) mock */
```
don't think we need these comments
edit: same for the one in the header

jori-nordic commented on tests/bluetooth/host/buf/cmake.txt:
```diff
@@ -15,6 +15,7 @@ FILE(GLOB module_mocks
   ${ZEPHYR_BASE}/tests/bluetooth/host/buf/mocks/net_buf.c
   ${ZEPHYR_BASE}/tests/bluetooth/host/buf/mocks/iso.c
   ${ZEPHYR_BASE}/tests/bluetooth/host/buf/mocks/hci_core.c
+  ${ZEPHYR_BASE}/tests/bluetooth/host/buf/mocks/buf_help_utils.c
```
I would rename to `net_buf_expects` or `net_buf_validator`, something more descriptive than utils.

ahmedmoheb-nordic replied:
I'd say to create a new `net_buf_expects` `.h` and `.c` pair to contain the current implementation and remove `buf_help_utils.c`
This because `buf_help_utils.h` contains some unrelated definitions to `net_buf`.

jori-nordic replied:
yes, I hadn't realized the pool getters were in there too

jori-nordic reviewed (changes requested):
An additional question: does the assert mock still work now that we're using FFF and that they're not enforcing one `expect` per fn call?

alwa-nordic commented on tests/bluetooth/host/buf/mocks/net_buf.h:
```diff
@@ -0,0 +1,25 @@
+#include <zephyr/fff.h>
+#include <zephyr/zephyr.h>
+
+DEFINE_FFF_GLOBALS;
```
Definitions in a header file. That should not be here.

alwa-nordic commented on tests/bluetooth/host/buf/mocks/net_buf.h:
```diff
@@ -0,0 +1,25 @@
+DEFINE_FFF_GLOBALS;
+
+/* List of fakes used by this unit tester */
+#define FFF_FAKES_LIST(FAKE)         \
```
I think this needs to be namespaced.

alwa-nordic commented on tests/bluetooth/host/buf/bt_buf_get_rx/src/main.c:
```diff
@@ -5,9 +5,9 @@
-#include <zephyr/bluetooth/hci.h>
 #include <zephyr/bluetooth/buf.h>
 #include "kconfig.h"
+#include "net_buf.h"
```
We should refer to this at least as `mocks/net_buf.h` for clarity.

alwa-nordic commented on tests/bluetooth/host/buf/bt_buf_get_rx/src/main.c:
```diff
@@ -5,9 +5,9 @@
-#include <zephyr/bluetooth/hci.h>
 #include <zephyr/bluetooth/buf.h>
 #include "kconfig.h"
+#include "net_buf.h"
```
What criteria do you use to select between `< >` and `" "` includes?

ahmedmoheb-nordic replied:
Just to refer that the file not part of a library as if it is only local to the project.
Technically, the compiler will handle it as if it is `<>` as the header file doesn't reside in the same directory

alwa-nordic commented on tests/bluetooth/host/buf/bt_buf_get_rx/src/main.c:
```diff
@@ -38,13 +38,13 @@ void test_returns_null_type_bt_buf_evt(void)
 
 	returned_buf = bt_buf_get_rx(BT_BUF_EVT, timeout);
 
+	validate_net_buf_alloc_called_behaviour(memory_pool, &timeout);
```
I wonder if this test is too restrictive. Why is it wrong for `bt_buf_get_rx` to result in multiple calls to `net_buf_alloc`? There must be a rationale for this restriction.

ahmedmoheb-nordic replied:
Without something describing the expected behavior, I'm extracting test cases from the current implementation.
Currently, a single call to `bt_buf_get_rx()` should result in a single call to `net_buf_alloc().`

When the implementation changes, the test should be adapted as well

ahmedmoheb-nordic committed:
tests: bluetooth: host: Code clean up, no logic change

This commit doesn't add any logical changes.
- add parent folder while including header files
- re-name test functions that validates expected behavior
- replace ZTest mocks with assert functions by FFF

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic commented:
All changes were cherry picked to [Add bluetooth host buf unit test](https://github.com/zephyrproject-rtos/zephyr/pull/47794)
